File: agencia-backend/src/main/java/com/agencia/agencia_backend/service/ProveedorService.java
package com.agencia.agencia_backend.service;

import com.agencia.agencia_backend.dto.CreateProveedorInput;
import com.agencia.agencia_backend.dto.UpdateProveedorInput;
import com.agencia.agencia_backend.model.Proveedor;
import com.agencia.agencia_backend.repository.ProveedorRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class ProveedorService {

    @Autowired
    private ProveedorRepository proveedorRepository;

    /**
     * Obtener todos los proveedores
     */
    public List<Proveedor> getAllProveedores() {
        return proveedorRepository.findAll();
    }

    /**
     * Obtener proveedor por ID
     */
    public Optional<Proveedor> getProveedorById(String id) {
        return proveedorRepository.findById(id);
    }

    /**
     * Obtener proveedores por tipo de servicio
     */
    public List<Proveedor> getProveedoresByTipoServicio(String tipoServicio) {
        return proveedorRepository.findByTipoServicio(tipoServicio);
    }

    /**
     * Buscar proveedores por nombre de empresa
     */
    public List<Proveedor> searchProveedores(String searchTerm) {
        List<Proveedor> proveedores = proveedorRepository.findAll();
        
        return proveedores.stream()
            .filter(proveedor -> {
                String nombreEmpresa = proveedor.getNombreEmpresa().toLowerCase();
                String contactoNombre = proveedor.getContactoNombre() != null ? 
                    proveedor.getContactoNombre().toLowerCase() : "";
                return nombreEmpresa.contains(searchTerm.toLowerCase()) || 
                       contactoNombre.contains(searchTerm.toLowerCase());
            })
            .toList();
    }

    /**
     * Crear nuevo proveedor
     */
    public Proveedor createProveedor(CreateProveedorInput input) {
        Proveedor proveedor = new Proveedor();
        proveedor.setNombreEmpresa(input.getNombreEmpresa());
        proveedor.setTipoServicio(input.getTipoServicio());
        proveedor.setContactoNombre(input.getContactoNombre());
        proveedor.setContactoEmail(input.getContactoEmail());
        proveedor.setContactoTelefono(input.getContactoTelefono());

        return proveedorRepository.save(proveedor);
    }

    /**
     * Actualizar proveedor existente
     */
    public Proveedor updateProveedor(String id, UpdateProveedorInput input) {
        Proveedor proveedor = proveedorRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Proveedor no encontrado con ID: " + id));

        // Actualizar solo los campos que no son null
        if (input.getNombreEmpresa() != null) {
            proveedor.setNombreEmpresa(input.getNombreEmpresa());
        }
        if (input.getTipoServicio() != null) {
            proveedor.setTipoServicio(input.getTipoServicio());
        }
        if (input.getContactoNombre() != null) {
            proveedor.setContactoNombre(input.getContactoNombre());
        }
        if (input.getContactoEmail() != null) {
            proveedor.setContactoEmail(input.getContactoEmail());
        }
        if (input.getContactoTelefono() != null) {
            proveedor.setContactoTelefono(input.getContactoTelefono());
        }

        return proveedorRepository.save(proveedor);
    }

    /**
     * Eliminar proveedor (hard delete)
     */
    public boolean deleteProveedor(String id) {
        Proveedor proveedor = proveedorRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Proveedor no encontrado con ID: " + id));

        proveedorRepository.delete(proveedor);
        return true;
    }
}

File: agencia-frontend/src/app/services/proveedor.service.ts
import { Injectable } from '@angular/core';
import { Apollo, gql } from 'apollo-angular';
import { Observable, map } from 'rxjs';
import { Proveedor, CreateProveedorInput, UpdateProveedorInput } from '../models/proveedor.model';

const GET_ALL_PROVEEDORES = gql`
  query GetAllProveedores {
    getAllProveedores {
      id
      nombreEmpresa
      tipoServicio
      contactoNombre
      contactoEmail
      contactoTelefono
    }
  }
`;

const GET_PROVEEDOR_BY_ID = gql`
  query GetProveedorById($id: ID!) {
    getProveedorById(id: $id) {
      id
      nombreEmpresa
      tipoServicio
      contactoNombre
      contactoEmail
      contactoTelefono
    }
  }
`;

const GET_PROVEEDORES_BY_TIPO_SERVICIO = gql`
  query GetProveedoresByTipoServicio($tipoServicio: String!) {
    getProveedoresByTipoServicio(tipoServicio: $tipoServicio) {
      id
      nombreEmpresa
      tipoServicio
      contactoNombre
      contactoEmail
      contactoTelefono
    }
  }
`;

const SEARCH_PROVEEDORES = gql`
  query SearchProveedores($searchTerm: String!) {
    searchProveedores(searchTerm: $searchTerm) {
      id
      nombreEmpresa
      tipoServicio
      contactoNombre
      contactoEmail
      contactoTelefono
    }
  }
`;

const CREATE_PROVEEDOR = gql`
  mutation CreateProveedor($input: CreateProveedorInput!) {
    createProveedor(input: $input) {
      id
      nombreEmpresa
      tipoServicio
      contactoNombre
      contactoEmail
      contactoTelefono
    }
  }
`;

const UPDATE_PROVEEDOR = gql`
  mutation UpdateProveedor($id: ID!, $input: UpdateProveedorInput!) {
    updateProveedor(id: $id, input: $input) {
      id
      nombreEmpresa
      tipoServicio
      contactoNombre
      contactoEmail
      contactoTelefono
    }
  }
`;

const DELETE_PROVEEDOR = gql`
  mutation DeleteProveedor($id: ID!) {
    deleteProveedor(id: $id)
  }
`;

@Injectable({
  providedIn: 'root'
})
export class ProveedorService {

  constructor(private apollo: Apollo) {}

  getAllProveedores(): Observable<Proveedor[]> {
    return this.apollo.query<{ getAllProveedores: Proveedor[] }>({
      query: GET_ALL_PROVEEDORES,
      fetchPolicy: 'network-only'
    }).pipe(
      map(result => result.data?.getAllProveedores || [])
    );
  }

  getProveedorById(id: string): Observable<Proveedor> {
    return this.apollo.query<{ getProveedorById: Proveedor }>({
      query: GET_PROVEEDOR_BY_ID,
      variables: { id },
      fetchPolicy: 'network-only'
    }).pipe(
      map(result => result.data!.getProveedorById)
    );
  }

  getProveedoresByTipoServicio(tipoServicio: string): Observable<Proveedor[]> {
    return this.apollo.query<{ getProveedoresByTipoServicio: Proveedor[] }>({
      query: GET_PROVEEDORES_BY_TIPO_SERVICIO,
      variables: { tipoServicio },
      fetchPolicy: 'network-only'
    }).pipe(
      map(result => result.data?.getProveedoresByTipoServicio || [])
    );
  }

  searchProveedores(searchTerm: string): Observable<Proveedor[]> {
    return this.apollo.query<{ searchProveedores: Proveedor[] }>({
      query: SEARCH_PROVEEDORES,
      variables: { searchTerm },
      fetchPolicy: 'network-only'
    }).pipe(
      map(result => result.data?.searchProveedores || [])
    );
  }

  createProveedor(input: CreateProveedorInput): Observable<Proveedor> {
    return this.apollo.mutate<{ createProveedor: Proveedor }>({
      mutation: CREATE_PROVEEDOR,
      variables: { input },
      refetchQueries: [{ query: GET_ALL_PROVEEDORES }]
    }).pipe(
      map(result => result.data!.createProveedor)
    );
  }

  updateProveedor(id: string, input: UpdateProveedorInput): Observable<Proveedor> {
    return this.apollo.mutate<{ updateProveedor: Proveedor }>({
      mutation: UPDATE_PROVEEDOR,
      variables: { id, input },
      refetchQueries: [{ query: GET_ALL_PROVEEDORES }]
    }).pipe(
      map(result => result.data!.updateProveedor)
    );
  }

  deleteProveedor(id: string): Observable<boolean> {
    return this.apollo.mutate<{ deleteProveedor: boolean }>({
      mutation: DELETE_PROVEEDOR,
      variables: { id },
      refetchQueries: [{ query: GET_ALL_PROVEEDORES }]
    }).pipe(
      map(result => result.data!.deleteProveedor)
    );
  }
}
